File: typescipt/existing/src/language/expressions/functions/lookupFunction.ts






namespace Lexy.Compiler.Language.Expressions.Functions;

internal class LookupFunction : ExpressionFunction, IHasNodeDependencies
{
   private const string FunctionHelp =
     "Arguments: LOOKUP(Table, lookUpValue, Table.SearchValueColumn, Table.ResultColumn)";

   public const string Name = "LOOKUP";

   private const int Arguments = 4;
   private const int ArgumentTable = 0;
   private const int ArgumentLookupValue = 1;
   private const int ArgumentSearchValueColumn = 2;
   private const int ArgumentResultColumn = 3;

   public string Table { get; }

   public Expression ValueExpression { get; }

   public MemberAccessLiteral ResultColumn { get; }
   public MemberAccessLiteral SearchValueColumn { get; }

   public VariableType ResultColumnType { get; private set; }
   public VariableType SearchValueColumnType { get; private set; }

   private LookupFunction(string tableType, Expression valueExpression,
     MemberAccessLiteral resultColumn, MemberAccessLiteral searchValueColumn,
     SourceReference tableNameArgumentReference)
     : base(tableNameArgumentReference)
   {
     Table = tableType ?? throw new ArgumentNullException(nameof(tableType));
     ValueExpression = valueExpression ?? throw new ArgumentNullException(nameof(valueExpression));
     ResultColumn = resultColumn ?? throw new ArgumentNullException(nameof(resultColumn));
     SearchValueColumn = searchValueColumn ?? throw new ArgumentNullException(nameof(searchValueColumn));
   }

   public IEnumerable<IRootNode> GetDependencies(RootNodeList rootNodeList)
   {
     var table = rootNodeList.GetTable(Table);
     if (table ! null) yield return table;
   }

   public static ParseExpressionFunctionsResult Parse(string name, SourceReference functionCallReference,
     IReadOnlyList<Expression> arguments)
   {
     if (arguments.Count ! Arguments)
       return ParseExpressionFunctionsResult.Failed($"Invalid number of arguments. {FunctionHelp}");

     if (!(arguments[ArgumentTable] is IdentifierExpression tableNameExpression))
       return ParseExpressionFunctionsResult.Failed(
         $"Invalid argument {ArgumentTable}. Should be valid table name. {FunctionHelp}");

     if (!(arguments[ArgumentSearchValueColumn] is MemberAccessExpression searchValueColumnHeader))
       return ParseExpressionFunctionsResult.Failed(
         $"Invalid argument {ArgumentSearchValueColumn}. Should be search column. {FunctionHelp}");

     if (!(arguments[ArgumentResultColumn] is MemberAccessExpression resultColumnExpression))
       return ParseExpressionFunctionsResult.Failed(
         $"Invalid argument {ArgumentResultColumn}. Should be result column. {FunctionHelp}");

     var tableName = tableNameExpression.Identifier;
     var valueExpression = arguments[ArgumentLookupValue];
     var searchValueColumn = searchValueColumnHeader.MemberAccessLiteral;
     var resultColumn = resultColumnExpression.MemberAccessLiteral;

     var lookupFunction = new LookupFunction(tableName, valueExpression, resultColumn, searchValueColumn,
       functionCallReference);
     return ParseExpressionFunctionsResult.Success(lookupFunction);
   }

   public override IEnumerable<INode> GetChildren()
   {
     yield return ValueExpression;
   }

   protected override void Validate(IValidationContext context)
   {
     ValidateColumn(context, ResultColumn, ArgumentResultColumn);
     ValidateColumn(context, SearchValueColumn, ArgumentSearchValueColumn);

     var tableType = context.RootNodes.GetTable(Table);
     if (tableType = null)
     {
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentTable}. Table name '{Table}' not found. {FunctionHelp}");
       return;
     }

     var resultColumnHeader = tableType.Header.Get(ResultColumn);
     if (resultColumnHeader = null)
     {
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentResultColumn}. Column name '{ResultColumn}' not found in table '{Table}'. {FunctionHelp}");
       return;
     }

     var searchColumnHeader = tableType.Header.Get(SearchValueColumn);
     if (searchColumnHeader = null)
     {
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentSearchValueColumn}. Column name '{SearchValueColumn}' not found in table '{Table}'. {FunctionHelp}");
       return;
     }

     var conditionValueType = ValueExpression.DeriveType(context);
     ResultColumnType = resultColumnHeader.Type.CreateVariableType(context);
     SearchValueColumnType = searchColumnHeader.Type.CreateVariableType(context);

     if (conditionValueType = null | !conditionValueType.Equals(SearchValueColumnType))
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentSearchValueColumn}. Column type '{SearchValueColumn}': '{SearchValueColumnType}' doesn't match condition type '{conditionValueType}'. {FunctionHelp}");
   }

   private void ValidateColumn(IValidationContext context, MemberAccessLiteral column, int index)
   {
     if (column.Parent ! Table)
       context.Logger.Fail(Reference,
         $"Invalid argument {index}. Result column table '{column.Parent}' should be table name '{Table}'");

     if (column.Parts.Length ! 2)
       context.Logger.Fail(Reference,
         $"Invalid argument {index}. Result column table '{column.Parent}' should be table name '{Table}'");
   }

   public override VariableType DeriveReturnType(IValidationContext context)
   {
     var tableType = context.RootNodes.GetTable(Table);
     var resultColumnHeader = tableType?.Header.Get(ResultColumn);

     return resultColumnHeader?.Type.CreateVariableType(context);
   }
}

File: src/Lexy.Compiler/Language/VariableTypes/Functions/LookUpFunction.cs
using System.Collections.Generic;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Language.Functions;
using Lexy.Compiler.Language.Tables;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.VariableTypes.Functions;

internal class LookUpFunction : TableFunctionReference
{
    private const string FunctionHelpValue =
        "Arguments: " +
           "Table.LookUp(lookUpValue, Table.ResultColumn) " +
        "or Table.LookUp(lookUpValue, Table.SearchColumn, Table.ResultColumn)" +
        "or Table.LookUp(discriminator, lookUpValue, Table.ResultColumn)" +
        "or Table.LookUp(discriminator, lookUpValue, Table.DiscriminatorColumn, Table.SearchColumn, Table.ResultColumn)";

    private record OverloadArguments(
        int? Discriminator,
        int LookUpValue,
        int? DiscriminatorColumnArgument,
        int? DefaultDiscriminatorColumn,
        int? SearchColumnArgument,
        int DefaultSearchColumn,
        int ResultColumnArgument);

    protected override string FunctionHelp => FunctionHelpValue;

    public const string Name = "LookUp";

    internal LookUpFunction(Table table): base(table)
    {
    }

    public override VariableType GetResultsType(IReadOnlyList<Expression> arguments)
    {
        var overloadArguments = GetArgumentColumns(null, arguments, null);

        var argument = arguments[overloadArguments.ResultColumnArgument];
        if (argument is not MemberAccessExpression columnExpression)
        {
            return null;
        }

        var column = Table.Header?.Get(columnExpression.VariablePath);
        return column?.Type.VariableType;
    }

    public override ValidateInstanceFunctionArgumentsResult ValidateArguments(IValidationContext context, IReadOnlyList<Expression> arguments,
        SourceReference reference)
    {
        if (!ValidateTable(context, reference)) return ValidateInstanceFunctionArgumentsResult.Failed();

        var overloadArguments = GetArgumentColumns(context, arguments, reference);
        var searchColumnHeader = GetColumn(context, arguments, overloadArguments.SearchColumnArgument, overloadArguments.DefaultSearchColumn, reference) ;
        var resultColumnHeader = GetColumn(context, arguments, overloadArguments.ResultColumnArgument, null, reference) ;

        if (!ValidateArguments(overloadArguments, searchColumnHeader, resultColumnHeader)) return ValidateInstanceFunctionArgumentsResult.Failed();

        ValidateColumnValueType(context, arguments, overloadArguments.LookUpValue, "Search", searchColumnHeader, reference);

        var discriminatorColumnHeader = ValidateDiscriminator(context, arguments, reference, overloadArguments);

        var result = new LookUpFunctionCall(
            Table.Name.Value,
            arguments[overloadArguments.LookUpValue],
            overloadArguments.Discriminator.HasValue ? arguments[overloadArguments.Discriminator.Value] : null,
            resultColumnHeader.Name,
            searchColumnHeader.Name,
            discriminatorColumnHeader?.Name);

        return ValidateInstanceFunctionArgumentsResult.Success(result);
    }

    private bool ValidateArguments(OverloadArguments overloadArguments, ColumnHeader searchColumnHeader, ColumnHeader resultColumnHeader)
    {
        if (searchColumnHeader == null || resultColumnHeader == null) return false;


        return true;
    }

    private ColumnHeader ValidateDiscriminator(IValidationContext context, IReadOnlyList<Expression> arguments,
        SourceReference reference, OverloadArguments overloadArguments)
    {
        if (overloadArguments.Discriminator == null) return null;

        var discriminatorColumnHeader = overloadArguments.DefaultDiscriminatorColumn != null
            ? GetColumn(context, arguments, overloadArguments.DiscriminatorColumnArgument, overloadArguments.DefaultDiscriminatorColumn, reference)
            : null;
        ValidateColumnValueType(context, arguments, overloadArguments.Discriminator.Value, "Discriminator", discriminatorColumnHeader, reference);

        return discriminatorColumnHeader;
    }

    private OverloadArguments GetArgumentColumns(IValidationContext context, IReadOnlyList<Expression> arguments, SourceReference reference)
    {
        switch (arguments.Count)
        {
            case 2:
                //"Table.LookUp(lookUpValue, Table.ResultColumn) " +
                return new OverloadArguments(null, 0, null, null, null, 0, 1);

            case 3:
                //"Table.LookUp(lookUpValue, Table.SearchColumn, Table.ResultColumn)"
                if (arguments[1] is MemberAccessExpression)
                {
                    return new OverloadArguments(null, 0, null, null, 1, 0, 2);
                }
                //"Table.LookUp(discriminator, lookUpValue, Table.ResultColumn)"
                return new OverloadArguments(0, 1, null, 0, null, 1, 2);

            case 5:
                //"Table.LookUp(discriminator, lookUpValue, Table.DiscriminatorColumn, Table.SearchColumn, Table.ResultColumn)";
                return new OverloadArguments(0, 1, 2, 0, 3, 1, 4);

            default:
                context?.Logger.Fail(reference, $"Invalid number of arguments. {FunctionHelpValue}");
                return null;
        }
    }
}